### PR TITLE
dealii: 9.5.0 and 9.5.1

### PR DESCRIPTION
**Pull request by luca-heltai**
I also cleaned a bit some parts of the package file.

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
Confirmed version sha256 and `spack spec` succeeds.  @jppelteret Do you want to review this PR as well?

**Comment by tldahlgren** [comment]
Looks like one of the specs failed (https://gitlab.spack.io/spack/spack/-/pipelines/441703):

```
...
--     COMPILE_DEFINITIONS: KOKKOS_DEPENDENCE
--     LINK_OPTIONS: -DKOKKOS_DEPENDENCE
-- 
-- DEAL_II_WITH_KOKKOS successfully set up with external dependencies.
-- 
-- Include /tmp/root/spack-stage/spack-stage-dealii-9.5.0-34xmwigxnn4ztarexfe4uxbli2cor57c/spack-src/cmake/configure/configure_40_cuda.cmake
-- DEAL_II_WITH_CUDA has unmet external dependencies.
CMake Error at cmake/configure/configure_40_cuda.cmake:119 (message):
  

  Could not find any suitable cuda library!

  deal.II can only be compiled with Cuda support if Kokkos was built with
  Cuda support!

  Please ensure that a cuda library is installed on your computer and deal.II
  is configured to use an external Kokkos installation.

Call Stack (most recent call first):
  /tmp/root/spack-stage/spack-stage-dealii-9.5.0-34xmwigxnn4ztarexfe4uxbli2cor57c/spack-build-34xmwig/CMakeFiles/CMakeTmp/evaluate_expression.tmp:1 (feature_CUDA_error_message)
  cmake/macros/macro_evaluate_expression.cmake:30 (include)
  cmake/macros/macro_configure_feature.cmake:275 (evaluate_expression)
  cmake/configure/configure_40_cuda.cmake:127 (configure_feature)
  cmake/macros/macro_verbose_include.cmake:19 (include)
  CMakeLists.txt:120 (verbose_include)


-- Configuring incomplete, errors occurred!
[1;31m==>[0m Error: ProcessError: Command exited with status 1:
...
```

**Comment by luca-heltai** [comment]
I have no way of testing if this works. @masterleinad, can you check if this works for you?

**Review by jppelteret — APPROVED** [review]
Thanks a lot for adding the new version and tidying up the package, @luca-heltai!

**Comment by tldahlgren** [comment]
Looks like CI is failing for one stack on `arborx` (https://gitlab.spack.io/spack/spack/-/jobs/7669468):

```
...
           h/linux-ubuntu20.04-x86_64_v3/gcc-11.1.0/kokkos-4.0.01-yg7j5hy4l6mug
           c6iyfa5ts6hktjvnhlg/lib/cmake/Kokkos (version "4.0.1")
     20    -- Could NOT find Kokkos_OPTIONS (missing: CUDA_LAMBDA)
  ^[[0;91m>> 21    CMake Error at /home/software/spack/__spack_path_placeholder__/__spa
           ck_path_placeholder__/__spack_path_placeholder__/__spack_path_placeh
           older__/__spack_path_placeholder__/__spack_path_placeholder__/__spac
           k_path_placeholder__/__spack_path_placeholder__/__spack_path_placeh/
           linux-ubuntu20.04-x86_64_v3/gcc-11.1.0/kokkos-4.0.01-yg7j5hy4l6mugc6
           iyfa5ts6hktjvnhlg/lib/cmake/Kokkos/KokkosConfigCommon.cmake:89 (mess
           age):^[[0m
     22      Kokkos does NOT provide all backends and/or architectures requeste
           d
     23    Call Stack (most recent call first):
     24      CMakeLists.txt:7 (kokkos_check)
     25
     26
     27    -- Configuring incomplete, errors occurred!

See build log for details:
  /tmp/root/spack-stage/spack-stage-arborx-1.4-2plgro5y5c5ps4wlcv3uoq4zts5zjaz4/spack-build-out.txt
  ```

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -179,6 +184,8 @@ class Dealii(CMakePackage, CudaPackage):
     # TODO: next line fixes concretization with petsc
     depends_on("hdf5+mpi+hl+fortran", when="+hdf5+mpi+petsc")
     depends_on("hdf5+mpi+hl", when="+hdf5+mpi~petsc")
+    depends_on("kokkos@3.7:", when="@9.5:+kokkos~trilinos")
+    depends_on("kokkos@3.7:+cuda+wrapper", when="@9.5:+kokkos~trilinos+cuda")
```
Comment (masterleinad) [inline var/spack/repos/builtin/packages/dealii/package.py]: We should probably do
```
    depends_on("kokkos+cuda_lambda", when="~trilinos+cuda")
```
as a workaround (see https://github.com/spack/spack/blob/3f4012cf441bcec98134332338d96bc2642015b1/var/spack/repos/builtin/packages/arborx/package.py#L82C1-L82C60) until we have https://github.com/spack/spack/pull/38957 or https://github.com/spack/spack/pull/38565 and lambda support is enabled by default.

Comment (luca-heltai) [reply var/spack/repos/builtin/packages/dealii/package.py]: Ok. I have no way to test this, so for me it is always a shot in the dark. I'll add your suggestions.

**Comment by masterleinad** [comment]
We are now getting
```
 Error: InstallError: Kokkos requires +wrapper when using +cudawithout clang
/builds/spack/spack/var/spack/repos/builtin/packages/kokkos/package.py:282, in cmake_args:
        279        if spec.satisfies("~wrapper+cuda") and not (
        280            spec.satisfies("%clang") or spec.satisfies("%cce")
        281        ):
  >>    282            raise InstallError("Kokkos requires +wrapper when using +cuda" "without clang")
        283
        284        options = [
        285            from_variant("CMAKE_POSITION_INDEPENDENT_CODE", "pic"),
```

**Comment by eugeneswalker** [comment]
Just pushed update to depend on `+cuda +cuda_lambda +wrapper` for `kokkos@3.7:`
* https://github.com/spack/spack/pull/38802/commits/2085e799eda74fe7bd000675ea76fbf31ae85642
```
    depends_on("kokkos@3.7:+cuda+cuda_lambda+wrapper", when="@9.5:+kokkos~trilinos+cuda")
```

**Comment by eugeneswalker** [comment]
`dealii +cuda` CI build is failing:
* [dealii/pvjff6n 9.5.0 gcc@=11.1.0 linux-ubuntu20.04-x86_64_v3 E4S](https://gitlab.spack.io/spack/spack/-/jobs/7734089)
```
...
  >> 777    FAILED: CMakeFiles/CheckCompilerSetupExec.dir/dummy.cpp.o
     778    /home/software/spack/__spack_path_placeholder__/__spack_path_placeh
            older__/__spack_path_placeholder__/__spack_path_placeholder__/__spa
            ck_path_placeholder__/__spack_path_placeholder__/__spack_path_place
            holder__/__spack_path_placeholder__/__spack_path_placeh/linux-ubunt
            u20.04-x86_64_v3/gcc-11.1.0/mpich-4.1.1-bkymmcmh6eev77ljrzo2on56hqa
            h53zv/bin/mpic++   -fopenmp-simd -pthread -std=c++17 -march=x86-64-
            v3 -mtune=generic -O0 -ggdb -Wa,--compress-debug-sections -extended
            -lambda -Wext-lambda-captures-this -arch=sm_80 -DKOKKOS_DEPENDENCE 
            -Wno-non-template-friend -o CMakeFiles/CheckCompilerSetupExec.dir/d
            ummy.cpp.o -c /tmp/root/spack-stage/spack-stage-dealii-9.5.0-pvjff6
            nymnjpambq4mhulrdoj4rkwsf7/spack-src/cmake/macros/check_compiler_se
            tup/dummy.cpp
  >> 779    g++: error: unrecognized command-line option '-Wext-lambda-captures
            -this'
  >> 780    g++: error: unrecognized command-line option '-arch=sm_80'
     781    ninja: build stopped: subcommand failed.
     782    
     783    
  >> 784    CMake Error at cmake/setup_finalize.cmake:126 (message):
     785    
     786    
     787        Configuration error: Cannot compile a test program with the fin
            al set of
     788        compiler and linker flags:
     789          CXX flags (DEBUG): -fopenmp-simd -pthread -std=c++17 -march=x
            86-64-v3 -mtune=generic -O0 -ggdb -Wa,--compress-debug-sections
     790          LD flags  (DEBUG): -rdynamic  -pthread -fopenmp -ggdb -Wl,--c
            ompress-debug-sections=zlib
```

It looks like some compilation commands are not being sent through `nvcc` when they should?

**Comment by masterleinad** [comment]
> It looks like some compilation commands are not being sent through `nvcc` when they should?

Yes, ideally `nvcc_wrapper` would be used as `CMAKE_CXX_COMPILER`.

**Comment by eugeneswalker** [comment]
> > It looks like some compilation commands are not being sent through `nvcc` when they should?
> 
> Yes, ideally `nvcc_wrapper` would be used as `CMAKE_CXX_COMPILER`.

What changes would we need to make here? Why did this work for past dealii version but not the new one?
* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py#L476-L501
```
        # MPI
        options.append(self.define_from_variant("DEAL_II_WITH_MPI", "mpi"))
        if "+mpi" in spec:
            options.extend(
                [
                    self.define("CMAKE_C_COMPILER", spec["mpi"].mpicc),
                    self.define("CMAKE_CXX_COMPILER", spec["mpi"].mpicxx),
                    self.define("CMAKE_Fortran_COMPILER", spec["mpi"].mpifc),
                    self.define("MPI_C_COMPILER", spec["mpi"].mpicc),
                    self.define("MPI_CXX_COMPILER", spec["mpi"].mpicxx),
                    self.define("MPI_Fortran_COMPILER", spec["mpi"].mpifc),
                ]
            )
            # FIXME: Fix issues with undefined references in MPI. e.g,
            # libmpi.so: undefined reference to `opal_memchecker_base_isaddressable'
            if "^openmpi" in spec:
                options.extend([self.define("MPI_CXX_LINK_FLAGS", "-lopen-pal")])
            if "+cuda" in spec:
                options.extend(
                    [
                        self.define(
                            "DEAL_II_MPI_WITH_CUDA_SUPPORT", spec["mpi"].satisfies("+cuda")
                        ),
                        self.define("CUDA_HOST_COMPILER", spec["mpi"].mpicxx),
                    ]
                )
```

**Comment by masterleinad** [comment]
> What changes would we need to make here? Why did this work for past dealii version but not the new one?

This is the first `deal.II` release to require `Kokkos` (and requires using `nvcc_wrapper` or `clang` when compiling with `Cuda` support).

**Comment by eugeneswalker** [comment]
> > What changes would we need to make here? Why did this work for past dealii version but not the new one?
> 
> This is the first `deal.II` release to require `Kokkos` (and requires using `nvcc_wrapper` or `clang` when compiling with `Cuda` support).

Can this PR be updated to make the changes @masterleinad suggests today? If this PR can be merged today we will be able to include it in our end-of-August E4S release.

**Comment by luca-heltai** [comment]
> > > What changes would we need to make here? Why did this work for past dealii version but not the new one?
> > 
> > 
> > This is the first `deal.II` release to require `Kokkos` (and requires using `nvcc_wrapper` or `clang` when compiling with `Cuda` support).
> 
> Can this PR be updated to make the changes @masterleinad suggests today? If this PR can be merged today we will be able to include it in our end-of-August E4S release.

I can try, but I will need some help. I have no access to any CUDA enabled machine. How do I set the CXX compilers to the nvcc_wrappers within spack?

**Comment by eugeneswalker** [comment]
> I can try, but I will need some help. I have no access to any CUDA enabled machine. How do I set the CXX compilers to the nvcc_wrappers within spack?

Trilinos sets it like this:
* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/trilinos/package.py#L537-L543
```
class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
...
    def setup_build_environment(self, env):
        spec = self.spec
        if "+cuda" in spec and "+wrapper" in spec:
            if "+mpi" in spec:
                env.set("OMPI_CXX", spec["kokkos-nvcc-wrapper"].kokkos_cxx)
                env.set("MPICH_CXX", spec["kokkos-nvcc-wrapper"].kokkos_cxx)
                env.set("MPICXX_CXX", spec["kokkos-nvcc-wrapper"].kokkos_cxx)
            else:
                env.set("CXX", spec["kokkos-nvcc-wrapper"].kokkos_cxx)
...
```

And `kokkos` looks like exposes `kokkos-nvcc-wrapper` like this:
* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/kokkos/package.py#L270-L274
```
class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
...
    def setup_dependent_package(self, module, dependent_spec):
        try:
            self.spec.kokkos_cxx = self.spec["kokkos-nvcc-wrapper"].kokkos_cxx
        except Exception:
            self.spec.kokkos_cxx = spack_cxx
...
```

I am testing changing this PR so that CMAKE_CXX_COMPILER is set to kokkos-nvcc-wrapper like this:
```
diff --git a/var/spack/repos/builtin/packages/dealii/package.py b/var/spack/repos/builtin/packages/dealii/package.py
index d8d780f182..3beb0a7399 100644
--- a/var/spack/repos/builtin/packages/dealii/package.py
+++ b/var/spack/repos/builtin/packages/dealii/package.py
@@ -508,7 +508,7 @@ def cmake_args(self):
             options.extend(
                 [
                     self.define("CMAKE_C_COMPILER", spec["mpi"].mpicc),
-                    self.define("CMAKE_CXX_COMPILER", spec["mpi"].mpicxx),
+                    self.define("CMAKE_CXX_COMPILER", spec["kokkos"].kokkos_cxx),
                     self.define("CMAKE_Fortran_COMPILER", spec["mpi"].mpifc),
                     self.define("MPI_C_COMPILER", spec["mpi"].mpicc),
                     self.define("MPI_CXX_COMPILER", spec["mpi"].mpicxx),
```

And getting much further in the build it seems, but now encountering this error:
```
==> Installing dealii-9.5.0-xq7scbitgeprjbmqep3kw6baipvnfchs [99/99]
==> No binary for dealii-9.5.0-xq7scbitgeprjbmqep3kw6baipvnfchs found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/a8/a81f41565f0d3a22d491ee687957dd48053225da72e8d6d628d210358f4a0464.tar.gz
==> No patches needed for dealii
==> dealii: Executing phase: 'cmake'
==> dealii: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/ninja-1.11.1-v75oqzh4kbvabdhrkck344vnyg2l3y7o/bin/ninja' '-j32' '-v'
...
  >> 14695    FAILED: source/cgal/CMakeFiles/object_cgal_debug.dir/surface_mesh.cc.o
...
  >> 14698    /usr/include/c++/11/array(111): error: incomplete type is not allowed
     14699              detected during:
     14700                instantiation of class "std::array<_Tp, _Nm> [with _Tp=CGAL::Point_2<CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>>, _Nm=2UL]"
     14701    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/cgal-5.5.2-3ed4rewq3g5aw76oqtisfcy73ooydo2w/include/CGAL/Cartesian/Ray_2.h(34): here
     14702                instantiation of class "CGAL::RayC2<R_> [with R_=CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>]"
     14703    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/cgal-5.5.2-3ed4rewq3g5aw76oqtisfcy73ooydo2w/include/CGAL/Ray_2.h(31): here
     14704                instantiation of class "CGAL::Ray_2<R_> [with R_=CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>]"
...
  >> 14770    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/cgal-5.5.2-3ed4rewq3g5aw76oqtisfcy73ooydo2w/include/CGAL/Cartesian/Line_3.h(39): error: incomplete type is not allowed
     14771              detected during:
     14772                instantiation of class "CGAL::LineC3<R_>::Rep [with R_=CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>]"
     14773    (46): here
     14774                instantiation of class "CGAL::LineC3<R_> [with R_=CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>]"
     14775    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/cgal-5.5.2-3ed4rewq3g5aw76oqtisfcy73ooydo2w/include/CGAL/Line_3.h(30): here
     14776                instantiation of class "CGAL::Line_3<R_> [with R_=CGAL::Simple_cartesian<CGAL::Interval_nt_advanced>]"
...
```

Any idea what this is about?

**Comment by luca-heltai** [comment]
@fdrmrc, do you have an idea of what the problem could be?

**Comment by fdrmrc** [comment]
> @fdrmrc, do you have an idea of what the problem could be?

It seems that CGAL doesn't support (officially) nvcc: https://github.com/CGAL/cgal/issues/7174.

**Comment by balay** [comment]
Checking if the issues here are with the new releases - or other feature changes to the spec file in this PR

**Comment by masterleinad** [comment]
The latest failure seems to be `Trilinos`-related. https://gitlab.spack.io/spack/spack/-/jobs/8539610 has
```
Error: KeyError: 'none'
/builds/spack/spack/var/spack/repos/builtin/packages/trilinos/package.py:941, in cmake_args:
        938                    ]
        939                )
        940                arch_map = Kokkos.spack_cuda_arch_map
  >>    941                options.extend(
        942                    define("Kokkos_ARCH_" + arch_map[arch].upper(), True)
        943                    for arch in spec.variants["cuda_arch"].value
        944                )
[...]
```

**Comment by masterleinad** [comment]
Potentionally that build requested `Cuda` support but didn't provide a `Cuda` architecture?

**Comment by eugeneswalker** [comment]
> Potentionally that build requested `Cuda` support but didn't provide a `Cuda` architecture?

https://github.com/spack/spack/pull/38802/files#diff-e4e73dbf3fb1228b717e5f3adb3040fab4cdb18af2ba220c606dbee2eb42edf3R240

Shouldn't the `cuda_arch` specified for dealii be forwarded to `trilinos+cuda` here?

This could be forwarded properly like it is done with trilinos package.py here:
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/trilinos/package.py#L393-L396
```
...
    for a in CudaPackage.cuda_arch_values:
        arch_str = "+cuda cuda_arch=" + a
        kokkos_spec = "kokkos@4.1.00 " + arch_str
        depends_on(kokkos_spec, when="@14.4.0 +kokkos " + arch_str)
...
```
Obviously the above snippet would be adapted for specifying the trilinos dependency in dealii.

**Comment by masterleinad** [comment]
> Shouldn't the `cuda_arch` specified for dealii be forwarded to `trilinos+cuda` here?

`deal.II` shouldn't specify any architecture but just take whatever `Kokkos` was configured with but I guess the `spack` -way is doing what you suggest
```
for a in CudaPackage.cuda_arch_values:
    arch_str = "+cuda cuda_arch=" + a
    trilinos_spec = "trilinos " + arch_str
    depends_on(trilinos_spec, when="@9.5:+trilinos " + arch_str)
```
@luca-heltai Can you try that? I don't think I can push to your branch.

**Pull request closed without merging**